File: src/daytrace/normalize.py
from __future__ import annotations

from collections.abc import Callable, Iterable, Mapping
from datetime import timedelta, timezone
from urllib.parse import urlsplit

from daytrace.models import (
    ActivityRecord,
    DayWindow,
    DiagnosticCode,
    RawBucket,
    RawEvent,
    SourceKind,
)


ONE_SECOND = timedelta(seconds=1)
SUPPORTED_BUCKET_TYPES: Mapping[str, SourceKind] = {
    "currentwindow": SourceKind.WINDOW,
    "afkstatus": SourceKind.AFK,
    "app.editor.activity": SourceKind.EDITOR,
    "web.tab.current": SourceKind.BROWSER,
}


def _text(data: Mapping[str, object], key: str) -> str | None:
    value = data.get(key)
    return value if isinstance(value, str) and value else None


def _url_parts(data: Mapping[str, object]) -> tuple[str | None, str | None]:
    value = _text(data, "url")
    if not value:
        return None, None
    try:
        parsed = urlsplit(value)
        return parsed.hostname.lower() if parsed.hostname else None, parsed.path or None
    except ValueError:
        return None, None


def normalize_events(
    bucket: RawBucket,
    events: Iterable[RawEvent],
    window: DayWindow,
    warn: Callable[[DiagnosticCode], None],
) -> tuple[ActivityRecord, ...]:
    kind = SUPPORTED_BUCKET_TYPES[bucket.type]
    normalized: list[ActivityRecord] = []
    window_start = window.start.astimezone(timezone.utc)
    window_end = window.end.astimezone(timezone.utc)

    for event in events:
        if event.timestamp.tzinfo is None:
            warn(DiagnosticCode.NAIVE_TIMESTAMP)
            continue
        raw_start = event.timestamp.astimezone(timezone.utc)
        start = max(raw_start, window_start)
        end = min(raw_start + event.duration_seconds * ONE_SECOND, window_end)
        if end <= start:
            warn(DiagnosticCode.NON_POSITIVE_EVENT)
            continue
        url_host, url_path = _url_parts(event.data)
        normalized.append(
            ActivityRecord(
                event_id=event.id,
                bucket_id=bucket.id,
                kind=kind,
                start=start,
                end=end,
                app=_text(event.data, "app"),
                title=_text(event.data, "title"),
                project=_text(event.data, "project"),
                file=_text(event.data, "file"),
                url_host=url_host,
                url_path=url_path,
                language=_text(event.data, "language"),
                status=_text(event.data, "status"),
            )
        )

    return tuple(
        sorted(normalized, key=lambda item: (item.start, item.bucket_id, item.event_id))
    )

File: packages/daytrace-core/src/normalize.ts
import { DaytraceInputError } from "./models.js";
import type {
  ActivityRecord,
  DayWindow,
  DiagnosticCode,
  RawBucket,
  RawEvent,
  SourceKind,
} from "./models.js";
import { timestampMs } from "./time.js";

export const SUPPORTED_BUCKET_TYPES: Readonly<Record<string, SourceKind>> = {
  currentwindow: "current-window",
  afkstatus: "afk",
  "app.editor.activity": "editor",
  "web.tab.current": "browser",
};

const ZONED_TIMESTAMP = /(?:Z|[+-]\d{2}:?\d{2})$/i;

function text(data: Readonly<Record<string, unknown>>, key: string): string | undefined {
  const value = data[key];
  return typeof value === "string" && value.length > 0 ? value : undefined;
}

function urlParts(data: Readonly<Record<string, unknown>>): {
  readonly urlHost?: string;
  readonly urlPath?: string;
} {
  const value = text(data, "url");
  if (value === undefined) return {};
  try {
    const parsed = new URL(value);
    return {
      ...(parsed.hostname.length === 0 ? {} : { urlHost: parsed.hostname.toLocaleLowerCase() }),
      ...(parsed.pathname.length === 0 ? {} : { urlPath: parsed.pathname }),
    };
  } catch {
    return {};
  }
}

function compareRecords(left: ActivityRecord, right: ActivityRecord): number {
  return timestampMs(left.start) - timestampMs(right.start)
    || left.bucketId.localeCompare(right.bucketId)
    || left.eventId.localeCompare(right.eventId);
}

export function normalizeEvents(
  bucket: RawBucket,
  events: readonly RawEvent[],
  window: DayWindow,
  warn: (code: DiagnosticCode) => void,
): readonly ActivityRecord[] {
  const kind = SUPPORTED_BUCKET_TYPES[bucket.type];
  if (kind === undefined) {
    throw new DaytraceInputError("unsupported-bucket", "unsupported ActivityWatch bucket");
  }
  const windowStart = timestampMs(window.start);
  const windowEnd = timestampMs(window.end);
  const normalized: ActivityRecord[] = [];
  for (const event of events) {
    if (!ZONED_TIMESTAMP.test(event.timestamp)) {
      warn("naive-timestamp");
      continue;
    }
    let rawStart: number;
    try {
      rawStart = timestampMs(event.timestamp);
    } catch {
      warn("naive-timestamp");
      continue;
    }
    const start = Math.max(rawStart, windowStart);
    const end = Math.min(rawStart + event.durationSeconds * 1000, windowEnd);
    if (end <= start) {
      warn("non-positive-event");
      continue;
    }
    const parts = urlParts(event.data);
    const app = text(event.data, "app");
    const title = text(event.data, "title");
    const project = text(event.data, "project");
    const file = text(event.data, "file");
    const language = text(event.data, "language");
    const status = text(event.data, "status");
    normalized.push({
      eventId: event.id,
      bucketId: bucket.id,
      kind,
      start: new Date(start).toISOString(),
      end: new Date(end).toISOString(),
      ...(app === undefined ? {} : { app }),
      ...(title === undefined ? {} : { title }),
      ...(project === undefined ? {} : { project }),
      ...(file === undefined ? {} : { file }),
      ...parts,
      ...(language === undefined ? {} : { language }),
      ...(status === undefined ? {} : { status }),
    });
  }
  return normalized.sort(compareRecords);
}
